torchmetrics/image/map.py

```diff
@@ -0,0 +1,369 @@
+            coco_preds.createIndex()
+            coco_eval = COCOeval(coco_target, coco_preds, "bbox")
+            coco_eval.evaluate()
+            coco_eval.accumulate()
+            coco_eval.summarize()
+            stats = coco_eval.stats
+
+        new_average_precision = torch.cat(
+            (
+                self.average_precision,
+                torch.tensor(
+                    [stats[COCO_STATS_MAP_VALUE_INDEX]], dtype=torch.float, device=self.average_precision.device
+                ),
+            )
+        )
+        setattr(self, "average_precision", new_average_precision)
+
+        new_average_recall = torch.cat(
+            (
+                self.average_recall,
+                torch.tensor([stats[COCO_STATS_MAR_VALUE_INDEX]], dtype=torch.float, device=self.average_recall.device),
+            )
+        )
+        setattr(self, "average_recall", new_average_recall)
+
+        # if class mode is enabled, evaluate metrics per class
+        for class_id in range(self.num_classes):
+            coco_eval.params.catIds = [class_id]
+            with _hide_prints():
+                coco_eval.evaluate()
+                coco_eval.accumulate()
+                coco_eval.summarize()
+                stats = coco_eval.stats
+
+            current_value = getattr(self, f"ap_{class_id}")
+            class_map_value = torch.cat(
+                (
+                    current_value,
+                    torch.tensor([stats[COCO_STATS_MAP_VALUE_INDEX]], dtype=torch.float, device=current_value.device),
+                )
+            )
+            setattr(self, f"ap_{class_id}", class_map_value)
+
+            current_value = getattr(self, f"ar_{class_id}")
+            class_map_value = torch.cat(
+                (
+                    current_value,
+                    torch.tensor([stats[COCO_STATS_MAR_VALUE_INDEX]], dtype=torch.float, device=current_value.device),
+                )
+            )
+            setattr(self, f"ar_{class_id}", class_map_value)
+
+        map_per_class_value = [torch.mean(getattr(self, f"ap_{class_id}")) for class_id in range(self.num_classes)]
+        mar_per_class_value = [torch.mean(getattr(self, f"ar_{class_id}")) for class_id in range(self.num_classes)]
+        metrics = MAPMetricResults(
+            map_value=torch.mean(self.average_precision),
+            mar_value=torch.mean(self.average_recall),
+            map_per_class_value=map_per_class_value,
+            mar_per_class_value=mar_per_class_value,
+        )
+        return metrics
+
+    def _get_coco_format(self, is_pred: bool = False) -> Dict:
+        """Transforms and returns all cached targets or predictions in COCO format.
+
+        Format is defined at https://cocodataset.org/#format-data
+        """
+
+        images = []
+        annotations = []
+        annotation_id = 1  # has to start with 1, otherwise COCOEval results are wrong
+
+        for i, _ in enumerate(self.groundtruth_boxes):
```

Comment:
Have you considered changing the function so that instead of taking ``is_pred``, it would take the boxes and labels that should be converted:

    def _get_coco_format(self, boxes, labels, scores=None)

It would be called like this:

    coco_target.dataset = self._get_coco_format(self.groundtruth_boxes, self.groundtruth_labels)
    coco_preds.dataset = self._get_coco_format(self.detection_boxes, self.detection_labels, self.detection_scores)

Then this loop could be written in a somewhat more compact form:

    for image_id, (image_boxes, image_labels) in enumerate(zip(boxes, labels)):
        for annotation_id, (box, label) in enumerate(zip(image_boxes, image_labels):
            ...
            if scores is not None:
                annotation["score"] = scores[image_id][annotation_id].cpu()

Reply:
Nice and clean, I like it and refactored the method. Thanks

Reply:
Had to adjust the `annotation_id` a bit, as it has to be unique across all images.